Remora.Commands/Builders/CommandParameterBuilder.cs

```diff
@@ -0,0 +1,453 @@
+            optionAttribute.ShortName,
+            optionAttribute.LongName,
+            _name,
+            _parameterType,
+            _isOptional,
+            _defaultValue,
+            _attributes,
+            _conditions,
+            description
+        );
+
+        return newNamedParameter;
+    }
+
+    private IParameterShape CreateNamedSingleValueParameterShape
+    (
+        OptionAttribute optionAttribute
+    )
+    {
+        var description = _description ?? Constants.DefaultDescription;
+
+        IParameterShape newNamedParameter = new NamedParameterShape
+        (
+            optionAttribute.ShortName,
+            optionAttribute.LongName,
+            _name,
+            _parameterType!,
+            _isOptional,
+            _defaultValue,
+            _attributes,
+            _conditions,
+            description
+        );
+
+        return newNamedParameter;
+    }
+
+    private IParameterShape CreateGreedyNamedSingleValueParameterShape
+    (
+        OptionAttribute optionAttribute
+    )
+    {
+        var description = _description ?? Constants.DefaultDescription;
+
+        IParameterShape newNamedParameter = new NamedGreedyParameterShape
+        (
+            optionAttribute.ShortName,
+            optionAttribute.LongName,
+            _name,
+            _parameterType!,
+            _isOptional,
+            _defaultValue,
+            _attributes,
+            _conditions,
+            description
+        );
+
+        return newNamedParameter;
+    }
+
+    private IParameterShape CreatePositionalParameterShape
+    (
+        RangeAttribute? rangeAttribute
+    )
+    {
+        var description = _description ?? Constants.DefaultDescription;
+        var isCollection = _parameterType!.IsSupportedCollection();
+
+        IParameterShape newPositionalParameter;
+        if (!isCollection)
+        {
+            var greedyAttribute = _attributes.OfType<GreedyAttribute>().SingleOrDefault();
+
+            newPositionalParameter = greedyAttribute is null
+                ? new PositionalParameterShape
+                  (
+                      _name,
+                      _parameterType!,
+                      _isOptional,
+                      _defaultValue,
+                      _attributes,
+                      _conditions,
+                      description
+                  )
+                : new PositionalGreedyParameterShape
+                  (
+                      _name,
+                      _parameterType!,
+                      _isOptional,
+                      _defaultValue,
+                      _attributes,
+                      _conditions,
+                      description
+                  );
+        }
+        else
+        {
+            newPositionalParameter = new PositionalCollectionParameterShape
+            (
+                rangeAttribute?.GetMin(),
+                rangeAttribute?.GetMax(),
+                _name,
+                _parameterType!,
```

Comment:
Assert instead of suppressing (`?? throw new ArgumentNullException()`) - fail fast, fail early. Same goes for the other ones. You can do this first thing in the method to avoid having to do it in multiple places. Check the other methods as well.